### ISSUE TITLE
Excessive MKL memory consumption with variable sized tensors

### ISSUE THREAD

**Issue by mavrov** [tensorflow/tensorflow#23455]
**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): No
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Linux Ubuntu 16.04
- Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device: N/A
- TensorFlow installed from (source or binary): Both source and binary
- TensorFlow version (use command below): 1.11.0
- Python version: 3.5.2
- Bazel version (if compiling from source): 0.15.0
- GCC/Compiler version (if compiling from source): gcc 5.4.0
- CUDA/cuDNN version: N/A
- GPU model and memory: N/A

**Describe the current behavior**
During inference of variable sized examples with Tensorflow MKL system memory consumption gradually increases.
If a server application is running long enough to go through a lot of examples of same rank but different shapes it would eventually run out of system memory.
After further investigation setting the [TF_MKL_OPTIMIZE_PRIMITVE_MEMUSE](https://github.com/tensorflow/tensorflow/blob/c19e29306ce1777456b2dbb3a14f511edf7883a8/tensorflow/core/util/mkl_util.h#L2041) to either 0 or 1 does not help.
Most MKL primitive factories ignore TF_MKL_OPTIMIZE_PRIMITVE_MEMUSE and even some that know about it may choose to ignore it based on other heuristics like [batches smaller than 32](https://github.com/tensorflow/tensorflow/commit/86d9ce130c5691cdba16024f7cc7987082acd294#diff-192dfeafbdd684934bdb0dfa8983674a) for example (I am not sure this is intended behavior).
These ["do_not_cache" heuristics](https://github.com/tensorflow/tensorflow/blob/c19e29306ce1777456b2dbb3a14f511edf7883a8/tensorflow/core/kernels/mkl_conv_ops.cc#L902
) on the other hand don't seem to follow the described logic in the comments above replacing "or" for "and".

**Describe the expected behavior**
There should be a way to disable MKL primitive memory reuse globally with an environment variable be it TF_MKL_OPTIMIZE_PRIMITVE_MEMUSE or better yet something else.
Primitive memory reuse does not play nice with unknown tensor sizes as most of the time there are no cache hits and allocated memory is simply piled up in the cache.

**Code to reproduce the issue**
I've attached a small example.
[min-leak-example.txt](https://github.com/tensorflow/tensorflow/files/2542973/min-leak-example.txt)

**Other info / logs**
I've attached Valgrind results of the example running on TF version 1.11.0 with and without MKL.
[min-leak-example-mkl.txt](https://github.com/tensorflow/tensorflow/files/2542977/min-leak-example-mkl.txt)
[min-leak-example-no-mkl.txt](https://github.com/tensorflow/tensorflow/files/2542978/min-leak-example-no-mkl.txt)

I've been working on a solution I can propose.
Should you find it reasonable I've extended the solution to full refactoring of the MKL primitive factories.

**Comment by gzmkl** [comment]
Just for trouble-shooting purpose, please experiment by hard-coding 
 do_not_cache = true;     (comment out existing logic). 

My initial feeling is that more and more MKL primitive caching due to new op input dimensions (thus new keys), so that the memory usage is going higher and higher (that is, the feeling of leakage). 

Of cause, this is just a guess. To confirm it, do the hard-coding experiment on all do_not_cache 
occurrences (in files of mkl*.cc in folder tensorflow/core/kernels). 

Thanks!

**Comment by vignesh-irtt** [comment]
> Your intuition is correct! I can confirm that setting do_not_cache = true and manually deleting the returned MklPrimitive pointer at the end of the scope works perfectly, in fact that's what I did for TF 1.9.
> 
> Otherwise the problem indeed looks like a memory leak but in fact a pointer to each allocated primitive is stored in static thread local hash maps. Unfortunately these pointers never get deleted.
> 
> Thank you for the suggestion and the quick response!

Hi, 
     i've tried ur idea of making do_not_cache=true, but memory leak still exists for me, can u plz give me ur tf-mkl wheel or ur repo

**Comment by mavrov** [comment]
The fork I am working on: [https://github.com/hyperscience/tensorflow](https://github.com/hyperscience/tensorflow)
You can start with this branch: [https://github.com/hyperscience/tensorflow/tree/ivan/mkl-variable-input-size-fix](https://github.com/hyperscience/tensorflow/tree/ivan/mkl-variable-input-size-fix)
This is further refactoring on the above branch: [https://github.com/hyperscience/tensorflow/tree/ivan/mkl-variable-input-size-fix-refactor](https://github.com/hyperscience/tensorflow/tree/ivan/mkl-variable-input-size-fix-refactor)
We have a patched r1.9 that works for us. The fixes are cherry picked in the downstream r1.11 and r1.12 branches.

**Comment by mavrov** [comment]
The problem is that there are a lot of primitive reuse cases and not all factories even have an option to turn caching off. In TF 1.9 turning the cache off for the convolution should be enough but in later versions the caching is becoming widely adopted in other operations as well and very few can be turned off. My proposed solution introduces a new environment variable with a default value that preserves current behaviour for backwards compatibility. It changes all factories so that they take this new variable into consideration when deciding to cache.

**Comment by vignesh-irtt** [comment]
> The problem is that there are a lot of primitive reuse cases and not all factories even have an option to turn caching off. In TF 1.9 turning the cache off for the convolution should be enough but in later versions the caching is becoming widely adopted in other operations as well and very few can be turned off. My proposed solution introduces a new environment variable with a default value that preserves current behaviour for backwards compatibility. It changes all factories so that they take this new variable into consideration when deciding to cache.

That's a nice idea because for training caching will help but for deploying caching will make the system's memory bloated which we currently experience.Thank You very much for ur response.

**Comment by vignesh-irtt** [comment]
> The problem is that there are a lot of primitive reuse cases and not all factories even have an option to turn caching off. In TF 1.9 turning the cache off for the convolution should be enough but in later versions the caching is becoming widely adopted in other operations as well and very few can be turned off. My proposed solution introduces a new environment variable with a default value that preserves current behaviour for backwards compatibility. It changes all factories so that they take this new variable into consideration when deciding to cache.

We tried ur branch r1.9 still we have the memory leak but it is reduced to **1/4th** previously we had and https://github.com/hyperscience/tensorflow/tree/ivan/mkl-variable-input-size-fix-refactor branch leaks more memory as normal tensorflow-mkl

**Comment by mavrov** [comment]
My r1.9 branch does not use the proposed solution. It is a quick and dirty patch that solved our problems back then. I have explicitly checked for other possible leaks of the same kind and could not locate such at Tensorflow version 1.9. 

The proposed solution (with refactoring) resides in the branch you've pointed out but I do not recommend building from it as it is branched at an arbitrary point from master. The solution is cherry picked in r1.11 and r1.12 that are supposed to be stable release branches.

And yes more leaks are happening at an increasing number of locations where the MklPrimitiveFactory is introduced in later Tensorflow versions. My branch is not the reason for those leaks and certainly cannot cause more leaking at the said locations compared to the same Tensorflow version.

And as I said the fix is backwards compatible so by default it does nothing unless you set TF_MKL_REUSE_PRIMITIVE_MEMORY to 0.

I hope someone looks into the problem soon. The refactored solution is already hard to rebase onto master.

**Comment by vignesh-irtt** [comment]
Okay that's why we settled with r1.9 since you said that's the branch where they started caching.Thank you

**Comment by tensorflowbutler** [comment]
Nagging Assignee @TensorFlow-MKL: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by Galeos93** [comment]
> My r1.9 branch does not use the proposed solution. It is a quick and dirty patch that solved our problems back then. I have explicitly checked for other possible leaks of the same kind and could not locate such at Tensorflow version 1.9.
> 
> The proposed solution (with refactoring) resides in the branch you've pointed out but I do not recommend building from it as it is branched at an arbitrary point from master. The solution is cherry picked in r1.11 and r1.12 that are supposed to be stable release branches.
> 
> And yes more leaks are happening at an increasing number of locations where the MklPrimitiveFactory is introduced in later Tensorflow versions. My branch is not the reason for those leaks and certainly cannot cause more leaking at the said locations compared to the same Tensorflow version.
> 
> And as I said the fix is backwards compatible so by default it does nothing unless you set TF_MKL_REUSE_PRIMITIVE_MEMORY to 0.
> 
> I hope someone looks into the problem soon. The refactored solution is already hard to rebase onto master.

Hi mavrov, I am currently having the same problem with variable sized inputs and I would like to try out your solution. If I understood correctly, one possible solution is build from branch https://github.com/hyperscience/tensorflow/tree/ivan/mkl-variable-input-size-fix-refactor. However, you recommend not building from it. Then, do you recommend merging the aforementioned branch on r1.12? Do I need to set TF_MKL_REUSE_PRIMITIVE_MEMORY to 0 or this environment variable is not yet implemented? Thank you in advance.

I hope this is fixed too, since it is quite troublesome for object localization tasks, where input images do not have a fixed size.

**Comment by gzmkl** [comment]
Hi,

Public PR has been created here
https://github.com/tensorflow/tensorflow/pull/24482

Hopefully it will be approved and merged soon.

**Comment by TensorFlow-MKL** [comment]
we are still monitoring this PR. not approved yet

**Comment by shahsahilj** [comment]
@TensorFlow-MKL 
I am facing a similar issue. Have detailed the issue at: https://stackoverflow.com/questions/56639787/how-to-fix-memory-leak-with-variable-tensors-in-tensorflow-mkl-library

Any further updates regarding the same?

**Comment by shahsahilj** [comment]
I found out that tensorflow 1.14.1 has code for changing the LRU_cache size in tensorflow/core/util/mkl_util.h.
I tried to build tf using that, but I wanted to know if the same change can be made in the 1.13.1 Some computations in 1.14 are making my program slower and I would like to use the 1.13 version

**Comment by agramesh1** [comment]
@shahsahilj  Can you retry the TF 1.14 that was released yesterday and measure perf again?  Some commits made yesterday will address performance degradation. 
It should be possible to back port the fix for LRU cache to 1.13.1 if  you need to do that.

**Comment by shahsahilj** [comment]
@agramesh1 Okay, I will recompile tf 1.14. How do I check which instruction sets I should compile it for?

**Comment by nathan-greeneltch-intel** [comment]
@shahsahilj we have a description of building from source in our install guide: https://software.intel.com/en-us/articles/intel-optimization-for-tensorflow-installation-guide

**Comment by shahsahilj** [comment]
Thank you for that. I am going to try and compile using those configs

**Comment by shahsahilj** [comment]
Sorry for the delayed response but the latest build of TF did fix those issues. Thanks!

**Comment by preethivenkatesh** [comment]
Please close this issue if fixed

**Comment by alankongfq** [comment]
I encounter this issue when running a DeepLab V3+ inference job with variable input size such as 192 x 192 and small batch size of 8 etc. I tried setting the environ variable TF_MKL_REUSE_PRIMITIVE_MEMORY = 0 as well as setting MKL_DISABLE_FAST_MM = 1 as per the guide over here [https://software.intel.com/content/www/us/en/develop/documentation/mkl-linux-developer-guide/top/managing-performance-and-memory/using-memory-functions/avoiding-memory-leaks-in-intel-mkl.html](url) but it does not seems to have any effect of solving the memory leak issue. When I load a model of size 512 x 512 and batch size of 8. The inference goes smoothly with no error. Any suggestion?

**Comment by NeoZhangJianyu** [comment]
@alankongfq We will check it.
Question: 
1. Could you provide the cmd to install the Tensorflow 1.15.0?
2. How do you check the memory leak?

Thank you!

**Comment by gzmkl** [comment]
@NeoZhangJianyu  I used to write a script to run Conv2D op with different input sizes (2500 of them, say 32x32, 128x128, etc), and then repeatedly run it. I used "htop" to monitor memory usage. This script helped to fix one customer issue which was related to MKL DNN primitive caching.

**Comment by NeoZhangJianyu** [comment]
@alankongfq Could you provide more detail to reproduce this issue?

**Comment by alankongfq** [comment]
@NeoZhangJianyu 
1. I had installed Tensorflow 1.15.0 using pip install
2. I used htop or system monitor to observe the memory usage.
I had managed to partially solve my problem as I check with my colleague who trained the model and he had mentioned that he trained in Tensorflow 2.0 so I did a switch and upgrade to the Tensorflow version and memory issue seems to not be there anymore. 

However, I had encounter another issue where the prediction was ok in GPU but not in CPU with MKL DNN libraries. I had since switch to running my inference on GPU as a workaround for the time being.

I would not be able to provide any codes as its work related.

**Comment by NeoZhangJianyu** [comment]
@alankongfq Thank you!
I have reproduced the memory leak issue in TF 1.15.
I have feedback to dev team.

I will try it on TF 2.x

For the prediction issue, could you try it on TF-eigen release? It don't use MKLDNN lib.
Install it: 

`conda install tensorflow=1.15.0=eigen_py36hd3854b5_0`
